### PR TITLE
[microNPU] Add a pass to reorder copy and compute nodes

### PR DESCRIPTION
**Pull request by NicolaLancellotti**
This pr adds a pass to reorder Arm(R) Ethos(TM)-U copy and compute nodes in such a way that independent DMA copies, and computes happen in parallel.

cc @lhutton1 @ekalda @manupa-arm

### CODE REVIEW

**Review by ekalda — COMMENTED** [review]
Looks really good, @NicolaLancellotti! I'd be interested in seeing how that pass interacts with graphs that have mixture of operators with and without weights, e.g. it seems to me that when we have a graph that looks like 
`pooling -> copy -> copy -> conv2d -> copy -> copy -> depthwise2d` 
it will end up after this pass as 
`copy -> copy -> pooling -> copy -> copy -> conv2d ... ` 
I suppose that is intentional, that we start copying the conv2d weights in while the MAC engine is crunching the pooling? Maybe it's worth adding a test that exercises that kind of mixture of ops?

File: src/tir/contrib/ethosu/passes.cc [hunk 1]
```diff
@@ -110,6 +110,71 @@ tvm::transform::Pass HoistAllocates() {
...
+      return StmtExprMutator::VisitStmt_(op);
+    }
+
+    auto seq_stmt{GetRef<SeqStmt>(op)};
+    std::vector<Stmt> new_seq(seq_stmt->size());
+    std::copy(seq_stmt->seq.begin(), seq_stmt->seq.end(), new_seq.begin());
+    bool previous_stmt_is_copy{true};  // Do not move the first stmt if it is a copy
+
+    for (size_t i{}; i < seq_stmt->size(); ++i) {
+      auto stmt{seq_stmt[i]};
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc]: Done.

File: src/tir/contrib/ethosu/passes.cc [hunk 2]
```diff
@@ -110,6 +110,71 @@ tvm::transform::Pass HoistAllocates() {
...
+      auto args{stmt.as<EvaluateNode>()->value.as<CallNode>()->args};
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc]: Done.

File: tests/python/contrib/test_ethosu/test_copy_compute_reordering.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            T.evaluate(T.call_extern("ethosu_copy", buffer6[0], 48, p6[0], dtype="handle"))
+            T.evaluate(T.call_extern("ethosu_conv2d", "int8", 4, 4, 4, 4, 0, 4, buffer1[0], 0, 0, 0, T.float32(0.00392081), -128, "NHWC", 16, 4, 1, "int8", 4, 4, 4, 4, 0, 4, p4[0], 0, 0, 0, T.float32(0.00839574), -128, "NHCWB16", 64, 16, 1, 1, 1, 1, 1, 1, 1, p1[0], 48, 0, p2[0], 48, 0, 0, 0, 0, "TANH", 0, 0, "TFL", "NONE", 0, 0, 0, dtype="handle"))
+            T.evaluate(T.call_extern("ethosu_copy", buffer7[0], 256, p7[0], dtype="handle"))
+            T.evaluate(T.call_extern("ethosu_depthwise_conv2d", "int8", 4, 4, 4, 4, 0, 4, p4[0], 0, 0, 0, T.float32(0.0078125), 0, "NHCWB16", 64, 16, 1, "int8", 4, 4, 4, 4, 0, 4, buffer8[0], 0, 0, 0, T.float32(0.00372155), -128, "NHWC", 16, 4, 1, 1, 1, 1, 1, 1, 1, p5[0], 16, 0, p6[0], 48, 0, 0, 0, 0, "TANH", 0, 0, "TFL", "NONE", 0, 0, 0, dtype="handle"))
+    # fmt: on
+
+    test_mod = CopyComputeReordering()(InputModule)
+    reference_mod = ReferenceModule
+    tvm.ir.assert_structural_equal(test_mod, reference_mod, True)
+
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply tests/python/contrib/test_ethosu/test_copy_compute_reordering.py:266]: Done.

File: src/tir/contrib/ethosu/passes.cc [hunk 4]
```diff
@@ -110,6 +110,71 @@ tvm::transform::Pass HoistAllocates() {
 
 TVM_REGISTER_GLOBAL("tir.contrib.ethos-u.HoistAllocates").set_body_typed(HoistAllocates);
 
+/*!
+ * \brief Reorders copy and compute nodes in such a way that independent DMA copies,
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc:114]: Done.

**Comment by NicolaLancellotti** [comment]
> I'd be interested in seeing how that pass interacts with graphs that have mixture of operators with and without weights, e.g. it seems to me that when we have a graph that looks like
> `pooling -> copy -> copy -> conv2d -> copy -> copy -> depthwise2d`
> it will end up after this pass as
> `copy -> copy -> pooling -> copy -> copy -> conv2d ... `
> I suppose that is intentional, that we start copying the conv2d weights in while the MAC engine is crunching the pooling? Maybe it's worth adding a test that exercises that kind of mixture of ops?

Yes, it is intentional, and the reordering is just what you said. I have added a test too.

**Comment by lhutton1** [comment]
Nice thanks @NicolaLancellotti! It looks like we need a re-trigger due to the linting issues CI had last week :)

**Review by lhutton1 — APPROVED** [review]
LGTM!

**Comment by lhutton1** [comment]
Looks like it now conflicts with #10344 :/

**Comment by NicolaLancellotti** [comment]
I have added a parameter to specify the maximum number of movements allowed for a copy.

File: src/tir/contrib/ethosu/passes.cc [hunk 5]
```diff
@@ -110,6 +110,98 @@ tvm::transform::Pass HoistAllocates() {
...
+    }
+
+    auto seq_stmt{GetRef<SeqStmt>(op)};
+    std::vector<Stmt> new_seq(seq_stmt->size());
+    std::copy(seq_stmt->seq.begin(), seq_stmt->seq.end(), new_seq.begin());
+
+    for (size_t index{}; index < new_seq.size(); ++index) {
+      for (int offset{}; offset < _max_copy_movements; ++offset) {
+        auto i{index - offset};
+        if (i > 0 && !stmt_is_copy(new_seq[i - 1]) && stmt_is_global_copy(new_seq[i])) {
```
Comment (manupa-arm) [inline src/tir/contrib/ethosu/passes.cc:146]: docs : please add some comment to explaint what is being done in the loop.

**Review by manupa-arm — CHANGES_REQUESTED** [review]
Thanks @NicolaLancellotti for the great work here.

I've left some comments to improve the code readability and a suggestion to allow PassContext to configure the pass.

File: src/tir/contrib/ethosu/passes.cc [hunk 6]
```diff
@@ -110,6 +110,98 @@ tvm::transform::Pass HoistAllocates() {
...
+    return args[0].as<StringImmNode>()->value == "ethosu_copy";
+  }
+
+  bool stmt_is_global_copy(Stmt stmt) {
+    auto args{get_stmt_args(stmt)};
+    return args[0].as<StringImmNode>()->value == "ethosu_copy" &&
+           args[3].as<BufferLoadNode>()->buffer.scope() == "global";
+  }
+
+  int _max_copy_movements;
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc:192]: Done.

File: src/tir/contrib/ethosu/passes.cc [hunk 7]
```diff
@@ -110,6 +110,98 @@ tvm::transform::Pass HoistAllocates() {
...
+/*!
+ * \brief A pass to reorder copy and compute nodes in such a way that independent DMA copies,
+ * and computes happen in parallel.
+ *
+ * \param max_copy_movements: The maximum number of movements allowed for a copy.
+ * \return tvm::transform::Pass
+ */
+tvm::transform::Pass CopyComputeReordering(int max_copy_movements) {
+  auto pass_func = [=](PrimFunc f, IRModule mod, tvm::transform::PassContext ctx) {
+    ICHECK(mod->GetGlobalVars().size() == 1 && mod->ContainGlobalVar("main"))
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc:206]: Done.

File: src/tir/contrib/ethosu/passes.cc [hunk 8]
```diff
@@ -110,6 +110,98 @@ tvm::transform::Pass HoistAllocates() {
...
+    auto eval_node{stmt.as<EvaluateNode>()};
+    ICHECK(eval_node) << "Expected statement to be an evaluate node, but was "
+                      << stmt->GetTypeKey();
+    auto call_node{eval_node->value.as<CallNode>()};
+    ICHECK(call_node) << "Expected expression to be a call node, but was "
+                      << eval_node->value->GetTypeKey();
+    return call_node->args;
+  }
+
+  bool stmt_is_copy(Stmt stmt) {
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc]: Done.

File: src/tir/contrib/ethosu/passes.cc [hunk 9]
```diff
@@ -110,6 +110,98 @@ tvm::transform::Pass HoistAllocates() {
...
+        auto i{index - offset};
+        if (i > 0 && !stmt_is_copy(new_seq[i - 1]) && stmt_is_global_copy(new_seq[i])) {
+          std::swap(new_seq[i], new_seq[i - 1]);
+        } else {
+          break;
+        }
+      }
+    }
+
+    auto n{CopyOnWrite(op)};
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc]: Done.

File: src/tir/contrib/ethosu/passes.cc [hunk 10]
```diff
@@ -110,6 +110,98 @@ tvm::transform::Pass HoistAllocates() {
...
+                      << eval_node->value->GetTypeKey();
+    return call_node->args;
+  }
+
+  bool stmt_is_copy(Stmt stmt) {
+    auto args{get_stmt_args(stmt)};
+    return args[0].as<StringImmNode>()->value == "ethosu_copy";
+  }
+
+  bool stmt_is_global_copy(Stmt stmt) {
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc]: Done.

File: src/tir/contrib/ethosu/passes.cc (hunk 5 above)
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc]: Done.

File: src/tir/contrib/ethosu/passes.cc [hunk 11]
```diff
@@ -110,6 +110,98 @@ tvm::transform::Pass HoistAllocates() {
...
+    if (op->size() <= 1) {
+      return StmtExprMutator::VisitStmt_(op);
+    }
+
+    auto seq_stmt{GetRef<SeqStmt>(op)};
+    std::vector<Stmt> new_seq(seq_stmt->size());
+    std::copy(seq_stmt->seq.begin(), seq_stmt->seq.end(), new_seq.begin());
+
+    for (size_t index{}; index < new_seq.size(); ++index) {
+      for (int offset{}; offset < _max_copy_movements; ++offset) {
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc]: Done.

File: src/tir/contrib/ethosu/passes.cc [hunk 12]
```diff
@@ -110,6 +110,98 @@ tvm::transform::Pass HoistAllocates() {
...
+        auto i{index - offset};
```
[the remark this answers is not in the document]
Comment (NicolaLancellotti) [reply src/tir/contrib/ethosu/passes.cc]: I refactored the pass and I added the type for this variable.

**Review by manupa-arm — APPROVED** [review]
LGTM!

File: src/tir/contrib/ethosu/passes.cc [hunk 13]
```diff
@@ -27,7 +27,17 @@
...
 
+#include <algorithm>
+
 namespace tvm {
+
+/*!
+ * \brief The maximum number of movements allowed for a copy in the CopyComputeReordering pass.
+ */
+constexpr const char* kCopyComputeReorderingMaxCopyMovements =
+    "tir.copy_compute_reordering_max_copy_movements";
```
Comment (manupa-arm) [inline src/tir/contrib/ethosu/passes.cc:38]: Since this is not a generic tir pass, lets use tir.ethosu.* naming

**Review by manupa-arm — COMMENTED** [review]
Ah just found one thing that needs to be fixed.
Happy to take an immediate follow up. Let me know

**Pull request merged**